File: Week1_EngineeringConcepts/Code/2_FactoryMethodPatternExample/src/DocumentFactory.java


public class DocumentFactory {

    public static class WordDocument implements Document {

        @Override
        public void performOperation() {
            System.out.println("Performing Operations in Word Document");
        }
    }

    public static class PdfDocument implements Document {

        @Override
        public void performOperation() {
            System.out.println("Performing Operations in PDF Document");
        }
    }

    public static class ExcelDocument implements Document {

        @Override
        public void performOperation() {
            System.out.println("Performing Operations in Excel Document");
        }
    }

    public Document createDocument(String type) {
        return switch (type.toLowerCase()) {
            case "word" ->
                new WordDocument();
            case "pdf" ->
                new PdfDocument();
            case "excel" ->
                new ExcelDocument();
            default ->
                throw new IllegalArgumentException(type);
        };
    }
}

File: Week1_EngineeringConcepts/Code/2_FactoryMethodPatternExample/DocumentFactory.cs
using System;
public class DocumentFactory
{
    public class WordDocument : Document
    {
        public void PerformOperation()
        {
            Console.WriteLine("Performing Operations in Word Document");
        }
    }
    public class PdfDocument : Document
    {
        public void PerformOperation()
        {
            Console.WriteLine("Performing Operations in PDF Document");
        }
    }
    public class ExcelDocument : Document
    {
        public void PerformOperation()
        {
            Console.WriteLine("Performing Operations in Excel Document");
        }
    }
    public Document CreateDocument(string type)
    {
        switch (type.ToLower())
        {
            case "word":
                return new WordDocument();
            case "pdf":
                return new PdfDocument();
            case "excel":
                return new ExcelDocument();
            default:
                throw new ArgumentException($"Invalid document type: {type}");
        }
    }
}
